src/Worker/Grpc/GrpcEntityRunner.cs

```diff
@@ -51,24 +52,99 @@ public static class GrpcEntityRunner
+    /// </exception>
+    /// <exception cref="ArgumentException">
+    /// Thrown if <paramref name="encodedEntityRequest"/> contains invalid data.
+    /// </exception>
+    public static async Task<string> LoadAndRunAsync(
+        string encodedEntityRequest, ITaskEntity implementation, ExtendedSessionsCache? extendedSessionsCache, IServiceProvider? services = null)
     {
         Check.NotNullOrEmpty(encodedEntityRequest);
         Check.NotNull(implementation);
 
         P.EntityBatchRequest request = P.EntityBatchRequest.Parser.Base64Decode<P.EntityBatchRequest>(
-            encodedEntityRequest);
+            encodedEntityRequest);
+        Dictionary<string, object?> properties = request.Properties.ToDictionary(
+                pair => pair.Key,
+                pair => ProtoUtils.ConvertValueToObject(pair.Value));
 
         EntityBatchRequest batch = request.ToEntityBatchRequest();
         EntityId id = EntityId.FromString(batch.InstanceId!);
-        TaskName entityName = new(id.Name);
-
+        TaskName entityName = new(id.Name);
+
+        bool addToExtendedSessions = false;
+        bool stateCached = false;
+        GrpcInstanceRunnerUtils.ParseRequestPropertiesAndInitializeCache(
+            properties,
+            extendedSessionsCache,
+            out double extendedSessionIdleTimeoutInSeconds,
+            out bool isExtendedSession,
+            out bool entityStateIncluded,
+            out MemoryCache? extendedSessions);
+
+        if (isExtendedSession && extendedSessions != null)
+        {
+            addToExtendedSessions = true;
+
+            // If an entity state was provided, even if we already have one stored, we always want to use the provided state.
+            if (!entityStateIncluded && extendedSessions.TryGetValue(request.InstanceId, out string? entityState))
+            {
+                batch.EntityState = entityState;
+                stateCached = true;
+            }
+        }
+
+        if (!stateCached && !entityStateIncluded)
+        {
+            // No state was provided, and we do not have one cached, so we cannot execute the batch request.
+            return Convert.ToBase64String(new P.EntityBatchResult { RequiresState = true }.ToByteArray());
+        }
+
         DurableTaskShimFactory factory = services is null
             ? DurableTaskShimFactory.Default
-            : ActivatorUtilities.GetServiceOrCreateInstance<DurableTaskShimFactory>(services);
-
-        TaskEntity entity = factory.CreateEntity(entityName, implementation, id);
-        EntityBatchResult result = await entity.ExecuteOperationBatchAsync(batch);
-
+            : ActivatorUtilities.GetServiceOrCreateInstance<DurableTaskShimFactory>(services);
+
+        TaskEntity entity = factory.CreateEntity(entityName, implementation, id);
+        EntityBatchResult result = await entity.ExecuteOperationBatchAsync(batch);
+
+        if (addToExtendedSessions)
+        {
+            extendedSessions!.Set(

```

Comment:
I suggest adding a comment briefly explaining why it's safe to assume that `extendedSessions` is never null here.